File: backend/run_migrations.py
#!/usr/bin/env python
"""
MongoDB Migration Runner
Usage:
    python run_migrations.py migrate      # Apply all pending migrations
    python run_migrations.py rollback     # Rollback last migration
    python run_migrations.py status       # Show migration status
"""

import asyncio
import sys
import os
from dotenv import load_dotenv
from motor.motor_asyncio import AsyncIOMotorClient

# Load environment variables
load_dotenv()

from migrations.migration_manager import MigrationManager

async def main():
    # Get command
    command = sys.argv[1] if len(sys.argv) > 1 else "status"
    
    # Connect to database
    mongo_url = os.getenv("MONGO_URL", "mongodb://localhost:27017/noteshub")
    client = AsyncIOMotorClient(mongo_url)
    
    # Get database name from URL or use default
    db_name = "noteshub"
    if "/" in mongo_url:
        db_name = mongo_url.split("/")[-1].split("?")[0]
    
    db = client[db_name]
    
    # Create migration manager
    manager = MigrationManager(db)
    
    try:
        if command == "migrate":
            print("🔄 Applying migrations...")
            await manager.migrate()
            print("✅ Migrations completed successfully")
            
        elif command == "rollback":
            print("⏪ Rolling back last migration...")
            await manager.rollback()
            print("✅ Rollback completed successfully")
            
        elif command == "status":
            print("📊 Migration Status:")
            status = await manager.status()
            print(f"\nTotal migrations: {status['total_migrations']}")
            print(f"Applied: {status['applied_count']}")
            print(f"Pending: {status['pending_count']}")
            
            if status['applied_migrations']:
                print("\n✅ Applied migrations:")
                for m in status['applied_migrations']:
                    print(f"  - {m['version']}: {m['description']}")
            
            if status['pending_migrations']:
                print("\n⏳ Pending migrations:")
                for m in status['pending_migrations']:
                    print(f"  - {m['version']}: {m['description']}")
        
        else:
            print(f"Unknown command: {command}")
            print("Usage: python run_migrations.py [migrate|rollback|status]")
            sys.exit(1)
            
    except Exception as e:
        print(f"❌ Error: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)
    finally:
        client.close()


if __name__ == "__main__":
    asyncio.run(main())

File: scripts/run-migrations.ts
/**
 * Database Migration Runner
 * Runs all pending migrations against the database
 */
import { drizzle } from 'drizzle-orm/postgres-js';
import { migrate } from 'drizzle-orm/postgres-js/migrator';
import postgres from 'postgres';
import dotenv from 'dotenv';
import { resolve } from 'path';

// Load environment variables
dotenv.config();

const DATABASE_URL = process.env.DATABASE_URL;

if (!DATABASE_URL) {
  console.error('❌ DATABASE_URL environment variable is not set');
  process.exit(1);
}

async function runMigrations() {
  console.log('🔄 Starting database migrations...');
  console.log('📁 Migrations folder: ./migrations');
  
  // Create connection for migrations
  const migrationClient = postgres(DATABASE_URL, { max: 1 });
  
  try {
    console.log('🔌 Connecting to database...');
    
    // Run migrations
    const db = drizzle(migrationClient);
    await migrate(db, { migrationsFolder: resolve(__dirname, '../migrations') });
    
    console.log('✅ Migrations completed successfully!');
  } catch (error) {
    console.error('❌ Migration failed:', error);
    throw error;
  } finally {
    // Close connection
    await migrationClient.end();
    console.log('🔌 Database connection closed');
  }
}

// Run migrations
runMigrations()
  .then(() => {
    console.log('\n✨ All done!');
    process.exit(0);
  })
  .catch((error) => {
    console.error('\n💥 Migration process failed:', error);
    process.exit(1);
  });
